ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3BatchDeleteHandler.java

```diff
@@ -0,0 +1,172 @@
+import com.fasterxml.jackson.dataformat.xml.ser.ToXmlGenerator;
+
+import static com.github.ambry.frontend.FrontendUtils.*;
+import static com.github.ambry.frontend.s3.S3Constants.*;
+import static com.github.ambry.rest.RestUtils.*;
+
+public class S3BatchDeleteHandler extends S3BaseHandler<ReadableStreamChannel> {
+  private static DeleteBlobHandler deleteBlobHandler;
+  private final XmlMapper xmlMapper;
+  private FrontendMetrics metrics;
+
+  // Constructor
+  public S3BatchDeleteHandler(DeleteBlobHandler deleteBlobHandler, FrontendMetrics frontendMetrics) {
+    this.deleteBlobHandler = deleteBlobHandler;
+    this.metrics = frontendMetrics;
+    this.xmlMapper = new XmlMapper();
+  }
+
+  /**
+   * Handles the S3 request and constructs the response.
+   *
+   * @param restRequest         the {@link RestRequest} that contains the request headers and body.
+   * @param restResponseChannel the {@link RestResponseChannel} that contains the response headers and body.
+   * @param callback            the {@link Callback} to invoke when the response is ready (or if there is an exception).
+   * @throws RestServiceException exception when the processing fails
+   */
+  @Override
+  protected void doHandle(RestRequest restRequest, RestResponseChannel restResponseChannel,
+      Callback<ReadableStreamChannel> callback) throws RestServiceException {
+
+    // Create the channel to read the request body
+    RetainingAsyncWritableChannel channel = new RetainingAsyncWritableChannel();
+
+    // inject metrics for batch delete
+    RestRequestMetrics requestMetrics =
+        metrics.batchDeleteMetricsGroup.getRestRequestMetrics(restRequest.isSslUsed(), false);
+    restRequest.getMetricsTracker().injectMetrics(requestMetrics);
+
+    // Pass the callback to handle the response
+    restRequest.readInto(channel,
+        parseRequestBodyAndDeleteCallback(channel, restRequest, deleteBlobHandler, restResponseChannel, callback));
+  }
+
+  /**
+   * This method will read the request body, deserialize it, and trigger the batch delete process.
+   */
+  private Callback<Long> parseRequestBodyAndDeleteCallback(RetainingAsyncWritableChannel channel,
+      RestRequest restRequest, DeleteBlobHandler deleteBlobHandler, RestResponseChannel restResponseChannel,
+      Callback<ReadableStreamChannel> finalCallback) {
+
+    return buildCallback(metrics.s3BatchDeleteHandleMetrics, bytesRead -> {
+      if (bytesRead == 0) {
+        throw new RestServiceException("bytesRead is empty", RestServiceErrorCode.BadRequest);
+      }
+
+      S3MessagePayload.DeleteResult response = new S3MessagePayload.DeleteResult();
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      ReadableStreamChannel readableStreamChannel =
+          new ByteBufferReadableStreamChannel(ByteBuffer.wrap(outputStream.toByteArray()));
+      S3MessagePayload.S3BatchDeleteObjects deleteRequest = deserializeRequest(channel);
```

Comment:
something seems odd, how are u deserializing the request from user ? L97 creates an new output stream that has nothing to do with user request.